cpp/src_prims/sparse/selection.cuh

```diff
@@ -0,0 +1,180 @@
+  int i = threadIdx.x;
+
+  int idx = row * n_cols;
+  K *inKStart = inK + idx + i;
+  IndexType *inVStart = inV + idx + i;
+
+  // Whole warps must participate in the selection
+  int limit = faiss::gpu::utils::roundDown(n_cols, faiss::gpu::kWarpSize);
+
+  for (; i < limit; i += tpb) {
+    inKStart = inK + idx + i;
+    inVStart = inV + idx + i;
+
+    heap.add(*inKStart, *inVStart);
+  }
+
+  // Handle last remainder fraction of a warp of elements
+  if (i < n_cols) {
+    inKStart = inK + idx + i;
+    inVStart = inV + idx + i;
+    heap.addThreadQ(*inKStart, *inVStart);
+  }
+
+  heap.reduce();
+
+  for (int i = threadIdx.x; i < k; i += tpb) {
+    outK[row * k + i] = smemK[i];
+    outV[row * k + i] = smemV[i];
+  }
+}
+
+template <typename value_idx = int, typename value_t = float, int warp_q,
+          int thread_q>
+inline void select_k_impl(value_t *inK, value_idx *inV, size_t n_rows,
+                          size_t n_cols, value_t *outK, value_idx *outV,
+                          bool select_min, int k, cudaStream_t stream) {
+  auto grid = dim3(n_rows);
+
+  constexpr int n_threads = (warp_q <= 1024) ? 128 : 64;
+  auto block = dim3(n_threads);
+
+  auto kInit = select_min ? faiss::gpu::Limits<float>::getMax()
+                          : faiss::gpu::Limits<float>::getMin();
+  auto vInit = -1;
+  if (select_min) {
+    select_k_kernel<value_t, value_idx, false, warp_q, thread_q, n_threads>
+      <<<grid, block, 0, stream>>>(inK, inV, n_rows, n_cols, outK, outV, kInit,
+                                   vInit, k);
+  } else {
+    select_k_kernel<value_t, value_idx, true, warp_q, thread_q, n_threads>
+      <<<grid, block, 0, stream>>>(inK, inV, n_rows, n_cols, outK, outV, kInit,
+                                   vInit, k);
+  }
+  CUDA_CHECK(cudaPeekAtLastError());
+}
+
+/**
+ * @brief Select the k-nearest neighbors from dense
+ * distance and index matrices.
+ *
+ * @param[in] inK partitioned knn distance matrix
+ * @param[in] inV partitioned knn index matrix
+ * @param[in] n_rows number of rows in distance and index matrices
+ * @param[in] n_cols number of columns in distance and index matrices
+ * @param[out] outK merged knn distance matrix
+ * @param[out] outV merged knn index matrix
+ * @param[in] select_min whether to select the min or the max distances
+ * @param[in] k number of neighbors per partition (also number of merged neighbors)
+ * @param[in] stream CUDA stream to use
+ */
+template <typename value_idx = int, typename value_t = float>
+inline void select_k(value_t *inK, value_idx *inV, size_t n_rows, size_t n_cols,
+                     value_t *outK, value_idx *outV, bool select_min, int k,
+                     cudaStream_t stream) {
+  if (k == 1)
+    select_k_impl<value_idx, value_t, 1, 1>(inK, inV, n_rows, n_cols, outK,
+                                            outV, select_min, k, stream);
+  else if (k <= 32)
+    select_k_impl<value_idx, value_t, 32, 2>(inK, inV, n_rows, n_cols, outK,
```

Comment:
Hmm... any way to collapse these calls and parametrize the launch params as variables? Would probably be more readable.

Reply:
This one is challenging. I have tried in the past to remove the need for all these instantiations but the `BlockSelect` inside the `select_k_kernel` requires both of those values to be constants.